File: prefeditor/src/main/java/com/github/kazy1991/prefeditor/SchemaSpinnerAdapter.java
package com.github.kazy1991.prefeditor;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.List;

public class SchemaSpinnerAdapter extends ArrayAdapter<PrefItem> {

    private LayoutInflater inflater;

    public SchemaSpinnerAdapter(@NonNull Context context, @NonNull List<PrefItem> objects) {
        super(context, 0, objects);
        inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        PrefItem item = getItem(position);
        View view = inflater.inflate(R.layout.cell_spinner, parent, false);
        ((TextView) view.findViewById(R.id.text)).setText(item.normalizedName());
        return view;
    }

    @Override
    public View getDropDownView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        PrefItem item = getItem(position);
        View view = inflater.inflate(R.layout.cell_spinner, parent, false);
        ((TextView) view.findViewById(R.id.text)).setText(item.normalizedName());
        return view;
    }
}

File: prefeditor/src/main/java/com/github/kazy1991/prefeditor/view/spinner/adapter/SchemaSpinnerAdapter.kt
package com.github.kazy1991.prefeditor.view.spinner.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.github.kazy1991.prefeditor.R
import com.github.kazy1991.prefeditor.entity.SchemaItem
import java.util.*

class SchemaSpinnerAdapter @JvmOverloads constructor(context: Context, objects: List<SchemaItem> = ArrayList()) : ArrayAdapter<SchemaItem>(context, 0, objects) {

    private val inflater: LayoutInflater = LayoutInflater.from(context)

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return setupView(position, convertView, parent)
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        return setupView(position, convertView, parent)
    }

    private fun setupView(position: Int, convertView: View?, parent: ViewGroup): View {
        val item = getItem(position)
        convertView?.let {
            val holder = it.tag as ViewHolder
            holder.textView.text = item!!.normalizedName()
            return it
        }.run {
            val view = inflater.inflate(R.layout.cell_spinner, parent, false)
            val holder = ViewHolder(view)
            holder.textView.text = item!!.normalizedName()
            view.tag = holder
            return view
        }
    }

    class ViewHolder(view: View) {
        val textView by lazy { view.findViewById<TextView>(R.id.text)!! }
    }

}
